### PR TITLE
Mjr/missing attachments

### PR DESCRIPTION
**Pull request by mjriley**
## Summary
This was created for [SAAS-11693](https://dimagi-dev.atlassian.net/browse/SAAS-11693). A customer has reported missing image attachments on a fairly regular basis. While this was previously due to issues on the mobile side, based on the Sumo Logic logs, it appears HQ should have received the most recent occurrences. Unfortunately, I've been unable to reproduce the issue or find code paths that make it seem likely, so this ticket is meant to provide further context to pin down where this could occur.

This uses the existing (albeit seldom used) logging framework we already have in place to log what form submission receives and the issues it runs into. A typical form submission results in the following output:

```
INFO 2021-01-14 21:39:16,240 submission_post 18579 123145418190848 Received Form 4afacc01-4c61-493a-85ac-35d60abb82b5 with 2 attachments
INFO 2021-01-14 21:39:16,241 submission_post 18579 123145418190848 Form 4afacc01-4c61-493a-85ac-35d60abb82b5, Attachment 0: 1610660349341.jpg (29447 bytes)
INFO 2021-01-14 21:39:16,241 submission_post 18579 123145418190848 Form 4afacc01-4c61-493a-85ac-35d60abb82b5, Attachment 1: form.xml
INFO 2021-01-14 21:39:19,106 submission_post 18579 123145418190848 Finished normal processing for Form 4afacc01-4c61-493a-85ac-35d60abb82b5
INFO 2021-01-14 21:39:25,294 submission_post 18579 123145418190848 Processing form fb7116ff-506d-491b-b0e1-d19ad2e04efd as a device log
```

~~One thing to highlight would be that the few uses of the Django logger we currently have all use the format `logging.<level>()`, whereas I instantiated loggers at the top of the affected files first, which follows Django's examples (and what I believe to be best practice) [here](https://docs.djangoproject.com/en/3.1/topics/logging/#using-logging). Both currently output to the same destination, but module-specific loggers give us more flexibility with what messages get outputting where.~~

There is an issue with our logging configuration that prevents us from following the recommended way to use loggers above. Within the LOGGING configuration of `settings.py`, we have `disable_existing_loggers: True`, which currently disables loggers as I initially tried to implement them. I believe this is caused by django.setup executing multiple times, but changing this behavior might result in a deluge of additional log messages, so this should be done as part of a future ticket.

I am a little concerned about the overhead of adding extra messages to the log file, but I think the above should not make a noticeable impact.

## Product Description

This only adds logging changes, so should be invisible to the user.

## Safety Assurance

- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change
- [x] If QA is part of the safety story, the "Awaiting QA" label is used
- [x] I am certain that this PR will not introduce a regression for the reasons below

### Automated test coverage

This ticket only adds logging messages, which do not change the 'contract' of our code -- the code never promises to log specific messages, and once this issue is resolved, it may make sense to remove them.

### QA Plan
No QA should be needed

### Safety story
This was tested locally along form submission paths I was evaluating (normal, duplicate, and device log submissions). I also verified that logging worked properly when a lock was unavailable. The other paths are relying on visual inspection.

### Rollback instructions

- [x] This PR can be reverted after deploy with no further considerations

### CODE REVIEW

File: corehq/apps/receiverwrapper/views.py [hunk 1]
```diff
@@ -61,6 +62,8 @@
 PROFILE_LIMIT = os.getenv('COMMCARE_PROFILE_SUBMISSION_LIMIT')
 PROFILE_LIMIT = int(PROFILE_LIMIT) if PROFILE_LIMIT is not None else 1
 
+logger = logging.getLogger(__file__)
```
Comment (millerdev) [inline corehq/apps/receiverwrapper/views.py:65]: Is there a reason you are using `__file__` instead of `__name__` here? I have not seen that before.

```suggestion
logger = logging.getLogger(__name__)
```

File: corehq/apps/receiverwrapper/views.py [hunk 2]
```diff
@@ -138,6 +141,7 @@ def _process_form(request, domain, app_id, user_id, authenticated,
         try:
             result = submission_post.run()
         except XFormLockError as err:
+            logger.warning('Unable to get lock for form %s' % err)
```
Comment (millerdev) [inline corehq/apps/receiverwrapper/views.py:144]: nit: inconsistent with other logging calls in this PR.
```suggestion
            logger.warning('Unable to get lock for form %s', err)
```

File: corehq/form_processor/parsers/form.py [hunk 3]
```diff
@@ -15,6 +16,8 @@
 from couchforms.exceptions import MissingXMLNSError
 from dimagi.utils.couch import release_lock
 
+logger = logging.getLogger(__file__)
```
Comment (millerdev) [inline corehq/form_processor/parsers/form.py:19]:
```suggestion
logger = logging.getLogger(__name__)
```

File: corehq/form_processor/parsers/form.py [hunk 4]
```diff
@@ -147,12 +150,15 @@ def _handle_id_conflict(xform, domain):
...
     if interface.is_duplicate(conflict_id, domain):
         # It looks like a duplicate/edit in the same domain so pursue that workflow.
+        logger.info('Handling duplicate doc id %s for domain %s', conflict_id, domain)
         return _handle_duplicate(xform)
     else:
         # the same form was submitted to two domains, or a form was submitted with
         # an ID that belonged to a different doc type. these are likely developers
         # manually testing or broken API users. just resubmit with a generated ID.
-        return interface.assign_new_id(xform), None
+        interface.assign_new_id(xform)
```
Comment (millerdev) [inline corehq/form_processor/parsers/form.py:159]: This changed slightly from previous behavior: the value returned by this method call used to be returned, but now is ignored. Looking at the implementations of `assign_new_id()` I see it does not matter since the object passed in is the object returned, but noting in case that was not an intentional change.

File: corehq/form_processor/submission_post.py [hunk 5]
```diff
@@ -322,9 +330,29 @@ def run(self):
...
+        logger.info('Received Form %s with %d attachments',
+            form.form_id, len(form.attachments_list))
+
+        for index, attachment in enumerate(form.attachments_list):
+            attachment_msg = 'Form %s, Attachment %s: %s'
+            attachment_props = [form.form_id, index, attachment.name]
+
+            if attachment.is_image:
+                attachment_msg = attachment_msg + ' (%d bytes)'
+                attachment_props.append(attachment.raw_content.size)
```
Comment (millerdev) [inline corehq/form_processor/submission_post.py:352]: Given the implementation of [`attachment.content_length`](https://github.com/dimagi/commcare-hq/blob/dbd92bed8dafd6d91e320ffc1ed9fdbb9e18a1e5/corehq/form_processor/models.py#L103), the type of `attachment.raw_content` apparently can vary. Are you sure it always has a `size` attribute? Note that the docstring says you should not use `attachment.content_length`, however maybe we should change that if it's needed here?

**Review by millerdev — COMMENTED** [review]
Test failures look related.

```
ERROR: corehq.apps.callcenter.tests.test_indicators:CallCenterCaseSharingTest.test_cases_owned_by_group
----------------------------------------------------------------------
Traceback (most recent call last):
  ...
  File "/mnt/commcare-hq-ro/corehq/form_processor/submission_post.py", line 258, in run
    self._log_form_details(submitted_form)
  File "/mnt/commcare-hq-ro/corehq/form_processor/submission_post.py", line 344, in _log_form_details
    form.form_id, len(form.attachments_list))
AttributeError: 'XFormInstance' object has no attribute 'attachments_list'
```

Edit: removed risk comment since @dimagimon already flagged it.

File: corehq/form_processor/submission_post.py (hunk 5 above)
Comment (mjriley) [reply corehq/form_processor/submission_post.py:352]: Right -- I found that the hard way when I was trying to output the size on form.xml, so now it only outputs for images. I believe all images should have raw_content, but I can probably put in some safety rails. I'm a little scared that this caused build failures, so more defensive programming might be in order

File: corehq/form_processor/parsers/form.py (hunk 4 above)
Comment (mjriley) [reply corehq/form_processor/parsers/form.py:159]: Yeah, I looked into the method and made sure that I wasn't changing behavior, as the intention was only to record the different ids

File: corehq/apps/receiverwrapper/views.py (hunk 1 above)
Comment (mjriley) [reply corehq/apps/receiverwrapper/views.py:65]: Nope, oversight on my part. __file__ would be a mistake here, as it would cause installations to different directories to create different loggers.

**Comment by mjriley** [comment]
> Test failures look related.

Yeah. The good news is that the normal operation of the system seems to always insert an attachments iterable, whether attachments were included or not. However, I think it makes sense to be more defensive and create an empty iterable if that property doesn't exist, rather than patching the affected tests.

File: corehq/form_processor/parsers/form.py (hunk 4 above)
Comment (millerdev) [reply corehq/form_processor/parsers/form.py:159]: Not necessary in this PR, but it might be reasonable to refactor `assign_new_id()` to not return a value and change all the places where it's called to not use the return value.

File: corehq/ex-submodules/couchforms/models.py [hunk 6]
```diff
@@ -368,6 +368,7 @@ def from_xform_instance(cls, instance, error_message, with_new_id=False):
         instance.__class__ = XFormError
         instance.doc_type = 'XFormError'
         instance.problem = error_message
+        instance.orig_id = None
```
Comment (millerdev) [inline corehq/ex-submodules/couchforms/models.py:371]: Why is this needed? Given that you're checking for the attribute at

https://github.com/dimagi/commcare-hq/blob/17130ad420fad472c02172979c0268007bb4f1eb/corehq/form_processor/submission_post.py#L356

**Review by millerdev — APPROVED** [review]
Approved, although would prefer to see `XFormError.orig_id` initialization removed if it's not absolutely necessary.

File: corehq/ex-submodules/couchforms/models.py (hunk 6 above)
Comment (mjriley) [reply corehq/ex-submodules/couchforms/models.py:371]: This was the hack I mentioned. We're taking an `XFormInstance` and relabelling it as an `XFormError` without running the proper initialization. Surprisingly, this means that running `hasattr(form, 'orig_id')` fails.

File: corehq/form_processor/submission_post.py [hunk 7]
```diff
@@ -322,9 +330,35 @@ def run(self):
...
+            attachment_props = [form.form_id, index, name]
+
+            if hasattr(attachment, 'has_size') and attachment.has_size():
+                attachment_msg = attachment_msg + ' (%d bytes)'
+                attachment_props.append(attachment.raw_content.size)
+
+            logger.info(attachment_msg, *attachment_props)
+
+    def _log_form_completion(self, form, submission_type):
+        # Orig_id only exists on SQL forms
```
Comment (snopoke) [inline corehq/form_processor/submission_post.py:355]: This comment is a bit misleading:
```suggestion
        # `orig_id` doesn't exist on all couch forms, only XFormError and XFormDeprecated
```

Comment (mjriley) [reply corehq/form_processor/submission_post.py]: Thanks Simon! Fixed.
